src/bin/ion/commands/dump.rs

```diff
@@ -168,12 +171,16 @@ fn write_all_values<W: IonWriter>(reader: &mut Reader, writer: &mut W) -> IonRes
             StreamItem::Nothing => break,
         }
         if reader.depth() == 0 {
+            index += 1;
             values_since_flush += 1;
             if values_since_flush == FLUSH_EVERY_N {
                 writer.flush()?;
                 values_since_flush = 0;
             }
         }
+        if head_number != -1 && index == head_number {
```

Comment:
We initialed the `head_number`(we call it `all-values` now) as -1. The `head_number` won't be updated until we execute `ion beta head `. In this case so the condition of `index >= head_number` will be true while we use `dump`, and this will affect the functionality of` ion dump`. Please correct me if I miss out any other logic :)

Reply:
By "the latter clause" I mean only the second half of this condition.